documentation/gant.1
.TH gant 1 "2008-05-24" "Russel Winder"
.SH NAME
gant \- A build framework based on scripting Ant task using Groovy.

.SH SYNOPSIS
.B gant
[\fIoptions\fR] [\fItarget\fR [\fItarget2\fR [\fItarget3\fR] ...\fR]\fR]

.SH DESCRIPTION

.B gant
is a tool for working with Ant tasks but using Groovy as the scripting language rather than using XML as a
specification language.

By default it takes information from
.B build.gant
which describes the targets.

.TP
\fB \-c,\-\-usecache\fR
Whether to cache the generated class and perform modified checks on the file before re-compilation.
.TP
\fB \-d,\-\-debug\fR
Print debug levels of information.
.TP
\fB \-f,\-\-file <build\-file>\fR
Use the named build file instead of the default,
.BR build.gant .
.TP
\fB \-h,\-\-help\fR
Print out this message.
.TP
\fB \-l,\-\-gantlib <library>\fR
A directory that contains classes to be used as extra Gant modules,
.TP
\fB \-n,\-\-dry\-run \fR
Do not actually action any tasks.
.TP
\fB \-p,\-\-projecthelp\fR
Print out a list of the possible targets.
.TP
\fB \-q,\-\-quiet \fR
Do not print out much when executing.
.TP
\fB \-s,\-\-silent\fR
Print out nothing when executing.
.TP
\fB \-v,\-\-verbose\fR
Print lots of extra information.
.TP
\fB \-C, \-\-cachedir <cache-file>\fR
The directory where to cache generated classes to.
.TP
\fB \-D <name>=<value>\fR
Define <name> to have value <value>.
Creates a variable named <name> for use in the scripts and a property
named <name> for the Ant tasks.
.TP
\fB \-L,\-\-lib <path>\fR
Add a directory to search for jars and classes.
.TP
\fB \-P,\-\-classpath <path-list>\fR
Specify a path list to search for jars and classes.
.TP
\fB \-T,\-\-targets\fR
Print out a list of the possible targets.
.TP
\fB \-V,\-\-version\fR
Print the version number and exit.

.SH AUTHOR
Gant and this manpage written by Russel Winder <russel@winder.org.uk>.


releaseNotes.txt
Updated to Groovy 1.8.6 and 2.0.0-beta-2.


1.9.7
-----

No substantive changes to Gant, per se.  Groovy 1.7.10, 1.8.4 and 2.0.0-beta-1 are the ones built against. 

Various minor changes to the build framework and project files have been made.  Various launch scripts have
been re-synchronized with those of Groovy.  Various changes made to allow for using Java 7 as well as Java
6, but none in Gant itself only in the tests.

This release actually triggered by the release of Groovy 2.0.0-beta-1 since it is a revolutionary change to
the numbering of Groovy -- 1.9.x series has been dropped and replaced by the 2.0.x series, i.e. the changes
have been deemed a major change.  Which is probably right.


1.9.6
-----

No substantive changes to Gant, this release is made post release of Groovy 1.8.1 and 1.9.0-beta-1
principally to get a version of Gant compiled against he Groovy 1.9 series into the Maven repository.

GroovyDoc, JavaDoc, and source artefacts are now built and published.


1.9.5
-----

GroovyShell.parse(InputStream, String) has been removed in Groovy 1.9.x.  This introduces and
incompatibility with the 1.6.x API.  Drop Gant support for Groovy 1.6.9.

Switch to using Groovy 1.8.0 (instead of 1.7.10) as the bundled version of Groovy for the stand-alone Gant
distribution.


1.9.4
-----

Add Groovy 1.8 to the Groovy series supported by Gant. 

Groovy 1.6.9 is now used for the 1.6 series build of Gant and 1.7.8 for the 1.7 series build.

Deb files are now built as part of the release process.

Remove the Maven Ant task jar from the repository and the distribution.


1.9.3
-----

Updated the Groovy 1.7.x series to be 1.7.3.

Minor fixes to the exit code and other processing for the Windows launch scripts to keep in line with Groovy
ones.

Corrected Windows batch file processing to avoid assumptions made in startGroovy.bat.

This release is really being made because it transpires that the 1.9.2 release for the Groovy 1.7 series was
compiled against Groovy 1.7.1 which had a breaking API change that was reverted for Groovy 1.7.2 and thus
1.7.3.  The consequence is that the version 1.9.2 gant_groovy1.7 artefact is incompatible with Groovy 1.7.2
and 1.7.3. 


1.9.2
------

Various minor bug fixes and refactorings, including switching to Groovy 1.7.1 as the main version of Groovy,
JUnit 4.8.1 as the unit test framework, and Ant 1.8.0 as the principal version of Ant used.

Maven and Gant builds removed in favour of focusing solely on Gradle as the build for the Gant project.


1.9.1
-----

The change to the Gradle build tool for release of Gant 1.9.0 led to a an error in the stream processing of
the launch scripts which led to a failure to start Gant.  This release is to fix that problem.


1.9.0
-----

Added the ability to skip the tests when using the Maven target set.  Use -DskipTest=true on the command
line to stop the tests being compiled and run, otherwise they will be.

Change the way in which the final message is output when Gant is executed from the command line from being
hardwired to being the execution of a hook, terminateHook.  cf. GANT-104.

Altered the Gant Ant task so as to provide an inheritAll attribute.  cf. GANT-110.

Switched GantException from extending RuntimeException to extending org.apache.tools.ant.BuildException so
as to allow better consistency with use of Gant from Ant.  cf. GANT-111.

Switch from using Ant as a boostrap and Gant for other processing to using Gradle for all build issues
(except distribution release as there is still a problem there).

Updated to Groovy 1.7.0 for Groovy 1.7.x and Groovy 1.6.7 for Groovy 1.6.x.

Switch from using Ant as a boostrap and then Gant to manage the project to using Gradle for all build and
install activity.


1.8.1
-----

Upgrade to Ivy 2.1.0.

Upgrade to Groovy 1.7-beta-2


1.8.0
-----

Upgrade to using JUnit 4.7, in line with Groovy 1.7.x.

Added the ability for each target to have a per-target prehook and posthook, and for there to be a global
prehook and posthook -- which are the same for all targets.  The hooks can either be a closure or a list of
closures.  Global prehooks are executed before per-target prehooks, which are executed before the build
listeners are informed and the target body is executed.  Global posthooks are executed after the per-target
posthooks which are executed after the target body and the buildlisteners.  The logging of entry and exit to
each target is implemented as the default per-target prehook/posthook. cf.  GANT-90, GANT-99 and GANT-100.

Remove (finally) the name Ant from the binding.  The symbol has been deprecated for ages.

Fixed GANT-91 and GANT-93.

Added nestedJavacCompilerArgs property to the Maven target set to allow for nested compilerarg elements to
the nested javac element of a groovyc element.

Upgraded to using Groovy 1.6.5.


1.7.0
-----

Added the binding variable initiatingTarget which is set by the Gant infrastructure just prior to processing
each target, it allows the Gant script to know what the initiating target is.

Added properties gant.version (the version of Gant currently executing) and gant.file (the initial Gant file
being processed) to the binding to allow the script to process these data.

Upgrade to using Groovy 1.5.8 instead of 1.5.7 as the Groovy 1.5.x version.

Upgrade to using Groovy 1.6.4 instead of 1.6.0 as the Groovy 1.6.x version.

Amend the way symbolic links are processed in the launch scripts on Mac OS X to correct various faults.

Added features to solve GANT-44.  The binding variable 'targets' is now an entry in the binding and contains
the list of as yet uncompleted targets.  This is the target list from the command line with all the already
processed ones removed.  The list contents are mutable within the Gant script but the variable itself is
read-only and so cannot be changed to refer to another list.

Various names present in the binding have been made read only within Gant scripts: target, message, ant,

src/main/groovy/org/codehaus/gant/ant/Gant.java
//  Gant -- A Groovy way of scripting Ant tasks.
//
//  Copyright © 2008-9 Russel Winder
//
//  Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in
//  compliance with the License. You may obtain a copy of the License at
//
//    http://www.apache.org/licenses/LICENSE-2.0
//
//  Unless required by applicable law or agreed to in writing, software distributed under the License is
//  distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
//  implied. See the License for the specific language governing permissions and limitations under the
//  License.

package org.codehaus.gant.ant ;

import java.io.File ;

import java.util.ArrayList ;
import java.util.Enumeration;
import java.util.HashMap ;
import java.util.Hashtable;
import java.util.List ;
import java.util.Map ;

import org.apache.tools.ant.AntClassLoader ;
import org.apache.tools.ant.BuildException ;
import org.apache.tools.ant.BuildListener ;
import org.apache.tools.ant.MagicNames ;
import org.apache.tools.ant.Project ;
import org.apache.tools.ant.Task ;

import org.codehaus.gant.GantBinding ;
import org.codehaus.gant.GantBuilder ;

/**
 *  Execute a Gant script.
 *
 *  <p>This Ant task provides a Gant calling capability. The original intention behind this was to support
 *  continuous integration systems that do not directly support Gant but only Ant.  However it also allows
 *  for gradual evolution of an Ant build into a Gant build.</p>
 *
 *  <p>Possible attributes are:</p>
 *
 *  <ul>
 *    <li>file &ndash; the path of the Gant script to execute.</li>
 *    <li>target &ndash; the target to execute; must be a single target name.  For specifying than a
 *        single target, use nested gantTarget tags.</li>
 *  </ul>
 *
 *  <p>Both of these are optional.  The file 'build.gant' and the default target are used by default.  An
 *  error results if there is no default target and no target is specified.</p>
 *
 *  <p>Definitions, if needed, are specified using nested <code>definition</code> tags, one for each symbol
 *  to be defined.  Each <code>definition</code> tag takes a compulsory <code>name</code> attribute and an
 *  optional <code>value</code> attribute.</p>
 *
 * @author Russel Winder
 */
public class Gant extends Task {
  /**
   *  The path to the file to use to drive the Gant build.  The default is build.gant.  This path is
   *  relative to the basedir of the Ant project if it is set, or the directory in which the job was started
   *  if the basedir is not set.
   */
  private String file = "build.gant" ;
  /**
   *  Flag determining whether properties are inherited from the parent project.
   */
  private boolean inheritAll = false ;
  /**
   *  A class representing a nested definition tag.
   */
  public static final class Definition {
    private String name ;
    private String value ;
    public void setName ( final String s ) { name = s ; }
    public String getName ( ) { return name ; }
    public void setValue ( final String s ) { value = s ; }
    public String getValue ( ) { return value ; }
  }
  /**
   *  A list of definitions to be set in the Gant instance.
   */
  private final List<Definition> definitions = new ArrayList<Definition> ( ) ;
  /**
   *  A class representing a nested target tag.
   */
  public static final class GantTarget {
    private String value ;
    public void setValue ( final String s ) { value = s ; }
    public String getValue ( ) { return value ; }
  }
  /**
   *  A list of targets to be achieved by the Gant instance.
   */
  private final List<GantTarget> targets = new ArrayList<GantTarget> ( ) ;
  /**
   *  Set the name of the build file to use.  This path is relative to the basedir of the Ant project if it
   *  is set, or the directory in which the job was started if the basedir is not set.
   *
   *  @param f The name of the file to be used to drive the build.
   */
  public void setFile ( final String f ) { file = f ; }
  /**
   *  Set the target to be achieved.
   *
   *  @param t The target to achieve.
   */
  public void setTarget ( final String t ) {
    final GantTarget gt = new GantTarget ( ) ;
    gt.setValue ( t ) ;
    targets.add ( gt ) ;
  }
  /**
   *  Create a node to represent a nested <code>gantTarget</code> tag.
   *
   *  @return a new <code>GantTarget</code> instance ready for values to be added.
   */
  public GantTarget createGantTarget ( ) {
    final GantTarget gt = new GantTarget ( ) ;
    targets.add ( gt ) ;
    return gt ;
  }
  /**
   *  Create a node to represent a nested <code>definition</code> tag.
   *
   *  @return a new <code>Definition</code> instance ready for values to be added.
   */
  public Definition createDefinition ( ) {
    final Definition definition = new Definition ( ) ;
    definitions.add ( definition ) ;
    return definition ;
  }
  /**
   *  If true, pass all properties to the new Ant project.
   *
   *  @param value if true pass all properties to the new Ant project.
   */
  public void setInheritAll ( final boolean value ) { inheritAll = value ; }
  /**
   * Load the file and then execute it.
   */
  @Override public void execute ( ) throws BuildException {
    //
    //  At first it might seem appropriate to use the Project object from the calling Ant instance as the
    //  Project object used by the AntBuilder object and hence GantBuilder object associated with the Gant
    //  instance we are going to create here.  However, if we just use that Project object directly then
    //  there are problems with proper annotation of the lines of output, so it isn't really an option.
    //  Therefore create a new Project instance and set the things appropriately from the original Project
    //  object.
    //
    //  Issues driving things here are GANT-50 and GANT-80.  GANT-50 is about having the correct base
    //  directory for operations, GANT-80 is about ensuring that all output generation actually generated
    //  observable output.
    //
    //  NB As this class is called Gant, we have to use fully qualified name to get to the Gant main class.
    //
    final Project antProject =  getOwningTarget ( ).getProject ( ) ;
    final Project newProject = new Project ( ) ;
    newProject.init ( ) ;
    //  Deal with GANT-80 by getting all the the loggers from the Ant instance Project object and adding
    //  them to the new Project Object.  This was followed up by GANT-91 so the code was amended to copying
    //  over all listeners except the class loader if present.
    for ( final Object o : antProject.getBuildListeners ( ) ) {
      final BuildListener listener = (BuildListener) o ;
      if ( ! ( listener instanceof AntClassLoader ) ) { newProject.addBuildListener ( listener ) ; }
    }
    //  Deal with GANT-50 by getting the base directory from the Ant instance Project object and use it for
    //  the new Project object.  GANT-93 leads to change in the way the Gant file is extracted.
    newProject.setBaseDir ( antProject.getBaseDir ( ) ) ;
    //  Deal with GANT-110 by using the strategy proposed by Eric Van Dewoestine.
    if ( inheritAll ) { addAlmostAll ( newProject , antProject ) ; }
    final File gantFile = newProject.resolveFile( file ) ;
    if ( ! gantFile.exists ( ) ) { throw new BuildException ( "Gantfile does not exist." , getLocation ( ) ) ; }
    final GantBuilder ant = new GantBuilder ( newProject ) ;
    final Map<String,String> environmentParameter = new HashMap<String,String> ( ) ;
    environmentParameter.put ( "environment" , "environment" ) ;
    ant.invokeMethod ( "property" , new Object[] { environmentParameter } ) ;
    final GantBinding binding = new GantBinding ( ) ;
    binding.forcedSettingOfVariable ( "ant" , ant ) ;
    for ( final Definition definition : definitions ) {
      final Map<String,String> definitionParameter = new HashMap<String,String> ( ) ;
      definitionParameter.put ( "name" , definition.getName ( ) ) ;
      definitionParameter.put ( "value" , definition.getValue ( ) ) ;
      ant.invokeMethod ( "property" , new Object[] { definitionParameter } ) ;
    }
    final gant.Gant gant = new gant.Gant ( binding ) ;
    gant.loadScript ( gantFile ) ;
    final List<String> targetsAsStrings = new ArrayList<String> ( ) ;
    for ( final GantTarget g : targets ) { targetsAsStrings.add ( g.getValue ( ) ) ; }
    final int returnCode =  gant.processTargets ( targetsAsStrings ) ;
    if ( returnCode != 0 ) { throw new BuildException ( "Gant execution failed with return code " + returnCode + '.' , getLocation ( ) ) ; }
  }
  /**
   *  Copy all properties from the given project to the new project -- omitting those that have already been
   *  set in the new project as well as properties named basedir or ant.file. Inspired by the {@code
   *  org.apache.tools.ant.taskdefs.Ant} source.
   *
   *  @param newProject the {@code Project} to copy into.
   *  @param oldProject the {@code Project} to copy properties from.
   */
  //  Russel Winder rehacked the code provided by Eric Van Dewoestine.
  private void addAlmostAll ( final Project newProject , final Project oldProject ) {
    @SuppressWarnings ( "unchecked" ) final Hashtable<String,String> properties = oldProject.getProperties ( ) ;
    final Enumeration<String> e = properties.keys ( ) ;
    while ( e.hasMoreElements ( ) ) {
      final String key = e.nextElement ( ) ;
      if ( ! ( MagicNames.PROJECT_BASEDIR.equals ( key ) || MagicNames.ANT_FILE.equals ( key ) ) ) {
        if ( newProject.getProperty ( key ) == null ) { newProject.setNewProperty ( key , properties.get ( key ) ) ; }
      }
    }
  }
}
